### PR TITLE
Implement decoding the binary form

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

**Comment by Hywan** [comment]
Fix #5.

**Review by Hywan — APPROVED** [review]
That's an impressive PR. I reviewed it carefully against the `BINARY.md` and I've found no error.

File: src/binary/decode.rs [hunk 1]
```diff
@@ -0,0 +1,2179 @@
...
+    ///
+    /// Since the custom section's arenas and AST nodes typically live inside
+    /// the `cx`, this is usually some sort of typed ID pointing into an arena.
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
+    where
+        Cx: Ids,
+        R: Read;
```
Comment (alexcrichton) [inline src/binary/decode.rs]: FWIW the codegen and wonkiness of `Read` I've found is pretty unfortunate historically, could this perhaps just be `&mut &[u8]` or something like that?

File: src/binary/decode.rs [hunk 2]
```diff
@@ -0,0 +1,2179 @@
...
+        &mut self,
+        cx: &mut Cx,
+    ) -> Result<Option<<T as Decode>::Output>, failure::Error>
+    where
+        T: Decode,
+        Cx: Ids;
+    fn string(&mut self) -> Result<String, failure::Error>;
+}
+
+impl<R: Read> ReadExt for R {
```
Comment (alexcrichton) [inline src/binary/decode.rs]: I wonder if it'd be a bit cleaner to expose these methods as `Decode for TheType`? For example the `Vec` method could be `Decode for Vec<T>` which albeit not as generic may still suffice? `uleb` could be something like `Decode for u32` or something like that too.

File: src/binary/decode.rs [hunk 3]
```diff
@@ -0,0 +1,2179 @@
...
+/// an input stream.
+pub(crate) trait Decode {
+    /// The output of decoding `Self` from the input stream.
+    ///
+    /// Since the custom section's arenas and AST nodes typically live inside
+    /// the `cx`, this is usually some sort of typed ID pointing into an arena.
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
```
Comment (alexcrichton) [inline src/binary/decode.rs]: This signature is really throwing me for a spin relative to what I'd expect from a `Decode` I think. I sort of just assumed that this would return `failure::Result<Self>` and then got pretty confused at all the usage of `Ignore` and `Output = ()`. I found it somewhat difficult to see what's being inserted where. 

Would it be possible to remove the `Output` associated type and change this to just return `Self`? I think that might help making it a bit more clear for me and having one location to insert types, for example, vs multiple locations that insert types.

**Review by alexcrichton — COMMENTED** [review]
Looks great to me! I find it kind of hard to follow the `Decode` trait but it's not really the end of the world. 

I might recommend switching to `&[u8]` forcibly though to avoid monomorphizing this entire module each time it's used, and it should also make codegen easier in rustc since dealing with generic reads tends to involve a lot of code anyway.

The tests look quite extensive though so I'd be fine merging whenever!

**Comment by Hywan** [comment]
How do you “extract” the resulting `WebidlBindings` struct with `on_parse`? The following code does not compile:

```rust
let mut bindings;
config.on_parse(|module, ids| { bindings = decode(ids, &mut data); Ok(()) });
```

I use `on_parse` because `IndicesToIds` is required for `decode`, and this is the only place where an instance of `IndicesToIds` exists.

Edit: It's possible to do everything inside `on_parse`, but am I missing something in the API?

File: src/binary/decode.rs (hunk 2 above)
Comment (fitzgen) [reply src/binary/decode.rs]: This would work for `u32` but not for `Vec<T>` since we don't always want to allocate a vec of the things we deserialize. I'll play with it a bit.

File: src/binary/decode.rs (hunk 1 above)
Comment (fitzgen) [reply src/binary/decode.rs]: Yeah, I've seen this before too. I'll see what I can do.

File: src/binary/decode.rs (hunk 3 above)
Comment (fitzgen) [reply src/binary/decode.rs]: Unfortunately, `Self` as the return type doesn't work so well with the arena approach. I did remove the `Cx` type param and am looking into removing the `Read` param too, which should help clean things up.

Comment (fitzgen) [reply src/binary/decode.rs]: I investigated this a bit but the hard part comes from thigns that reference other things by index (like webidl types) where we haven't yet finished constructing a whole `WebidlTypes` and its arena. This is why we sometimes decode by mutating the under-construction custom section in the cx instead of always returning the decoded value. I don't see a nice/easy way to work around this.
